### PR TITLE
New Exercise: Scrabble Score

### PR DESCRIPTION
**Pull request by jvarness**
I haven't actually ran the presubmit and formatter check on this code yet. Just getting this out there for review.

### CODE REVIEW

**Comment by jvarness** [comment]
Hey, what's the best way to get the `config.json` generated for this?

**Comment by Stargator** [comment]
Hey @jvarness, thanks for coming over from dev.to!

[CONTRIBUTING](https://github.com/exercism/dart/blob/master/CONTRIBUTING.md) should have the relevant guidance.

I don't believe we re-generate the config.json. You would just need to update it with the information for the new exercise. Preferably at the bottom of the list. Guess for any values you may not be sure about and we can review them later.

**Comment by Stargator** [comment]
@jvarness In the README it suggests to use dart bin/create_exercise.dart to generate the test file and some related files.

You can run it with `dart bin/create_exercise.dart`.

Do comment out the following line from it so it overwrites your existing files and you can diff the changes and make any corrections you feel warranted.

```dart
  if (await exerciseDir.exists()) {
    stderr.write("$name already exist\n");
    exit(1);
  }
```

**Comment by Stargator** [comment]
> @Stargator I tried executing bin/create_exercise.dart, but all it did was get rid of all the work I have already done.

Do you have a solid Diff tool? Not sure what platform you're using, but after you run the script, use a Diff comparison tool like BeyondCompare to see the differences. Then you can manually make some corrections.

Stuff like adding the `skip` automatically and consistently generation of the README and other files would help. If you feel uncomfortable doing that, then I'll do it locally and just add comments for the suggested changes.

File: exercises/scrabble-score/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+- = `3 + 2 + 6 + 3`
+- = `5 + 9`
+- = 14
+
+## Extensions
+
+- You can play a double or a triple letter.
+- You can play a double or a triple word.
+
+## Source
```
Comment (Stargator) [inline exercises/scrabble-score/README.md]: Add the following above Source, including a line before and after:

```Markdown

To run the tests:

```sh
$ pub run test
```

For more detailed info about the Dart track see the [help page](http://exercism.io/languages/dart).

```

File: exercises/scrabble-score/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+Given a word, compute the scrabble score for that word.
```
Comment (Stargator) [inline exercises/scrabble-score/README.md]: Add the following as the first line, then add another line after it:

```Markdown
# Scrabble Score

```

File: exercises/scrabble-score/lib/scrabble_score.dart [hunk 3]
```diff
@@ -0,0 +1,3 @@
+class ScrabbleScore {
+  // your code here
```
Comment (Stargator) [inline exercises/scrabble-score/lib/scrabble_score.dart]: Change to `// Put your code here`

Add a line after the `}`

File: exercises/scrabble-score/test/scrabble_score_test.dart [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    group('should return a score of 0 for', () {
+      test('an empty string', () {
+        var word = new ScrabbleScore('');
+        expect(word.getScrabbleScore(), equals(0));
+      });
+
+      test('strings containing only whitespace characters', () {
+        var word = new ScrabbleScore(' \t\r\n');
+        expect(word.getScrabbleScore(), equals(0));
+      });
```
Comment (Stargator) [inline exercises/scrabble-score/test/scrabble_score_test.dart]: Add `skip: true` to all tests after the first.

Add `skip: false` to the first test.

File: exercises/scrabble-score/README.md (hunk 2 above)
Comment (jvarness) [reply exercises/scrabble-score/README.md]: That's been added.

File: exercises/scrabble-score/lib/scrabble_score.dart (hunk 3 above)
Comment (jvarness) [reply exercises/scrabble-score/lib/scrabble_score.dart]: Added

**Comment by jvarness** [comment]
@Stargator I'm going to probably end up rewriting the test cases to more closely match the canonical-data for the problem. I was trying to be cool and write all my own.

This is the first exercise I've tried to implement. Next time, I will use the exercise generator so that everything will be less painful to review.

**Comment by jvarness** [comment]
@Stargator I've updated the test cases to only include those specified in the canonical data.

**Comment by Stargator** [comment]
@jvarness No worries, thanks for adding a new exercise. It helps so i understand where some of the pain points might be for new contributors.

File: exercises/scrabble-score/lib/scrabble_score.dart (hunk 3 above)
Comment (jvarness) [reply exercises/scrabble-score/lib/scrabble_score.dart]: Sorry. Added `Put`

File: exercises/scrabble-score/test/scrabble_score_test.dart [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        var word = new ScrabbleScore('');
+        expect(word.getScrabbleScore(), equals(0));
+      }, skip: false);
+    });
+
+    group('should return the appropriate score for', () {
+
+      test('lowercase letters', () {
+        var word = new ScrabbleScore('a');
+        expect(word.getScrabbleScore(), equals(1));
```
Comment (Stargator) [inline exercises/scrabble-score/test/scrabble_score_test.dart]: Just noting that to fit the canonical-data.json's template the tests should look more like:

```dart
    test("lowercase letters", () {
      final int result = scrabbleScore.score("a");
      expect(result, equals(1));
    });
```

No constructor, just a method called score that is passed the input.

Comment (jvarness) [reply exercises/scrabble-score/test/scrabble_score_test.dart]: If that's the case, would it be more proper to utilize Dart's support for high-order functions? No classes, just functions?

**Comment by jvarness** [comment]
@Stargator I've updated the example and the test cases and example to support high-order functions. I've also rebased off of the current master and squashed my commits.

File: exercises/scrabble-score/lib/example.dart [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  'w': 4,
+  'y': 4,
+  'k': 5,
+  'j': 8,
+  'x': 8,
+  'z': 10,
+  'q': 10
+};
+
+int getScrabbleScore(String word) {
```
Comment (Stargator) [inline exercises/scrabble-score/lib/example.dart]: Can you rename the function to score? So it fits the property "property" that's defined in it.

File: exercises/scrabble-score/lib/example.dart [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  int val = 0;
+  if(word.length > 0) {
```
Comment (Stargator) [inline exercises/scrabble-score/lib/example.dart]: Run `dartfmt -i 0 -l 120 -w .` so the exercise is properly formatted.

File: exercises/scrabble-score/pubspec.yaml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+name: 'scrabble_score'
+dev_dependencies:
+  test: '<0.13.0'
```
Comment (Stargator) [inline exercises/scrabble-score/pubspec.yaml]: In the create_exercise_test.dart, extend the timeout to 2 minutes. As the tests are timing out on Travis CI.

Though the builds are not failing properly, so I'm going to make another ticket for it. So I'll have to double check the contents to ensure the results are correct.

File: exercises/scrabble-score/lib/example.dart (hunk 6 above)
Comment (jvarness) [reply exercises/scrabble-score/lib/example.dart]: Done

File: exercises/scrabble-score/lib/example.dart (hunk 7 above)
Comment (jvarness) [reply exercises/scrabble-score/lib/example.dart]: Done

File: exercises/scrabble-score/pubspec.yaml (hunk 8 above)
Comment (jvarness) [reply exercises/scrabble-score/pubspec.yaml]: I think you may have meant `exercises_test.dart`?

I changed the Duration to 120 seconds.

Comment (Stargator) [reply exercises/scrabble-score/pubspec.yaml]: I stand corrected, yep.

**Comment by jvarness** [comment]
@Stargator I believe I've covered everything.

**Comment by Stargator** [comment]
@jvarness Sorry for the back and forth, I thought I had already submitted the feedback about the null.

Thanks for your work and patience.

**Comment by jvarness** [comment]
@Stargator no problem. I changed the `config.json` file to have `null` and not `"null"`

**Comment by Stargator** [comment]
@jvarness Reviewing today, just gotta finish refactoring something on another project and then I'll run through your changes.

**Comment by Stargator** [comment]
Approved, just recommend moving:
>     {
>      "core": true,
>      "difficulty": 1,
>      "slug": "scrabble-score",
>      "topics": [
>        "control_flow_conditionals",
>        "control_flow_loops",
>        "equality",
>        "strings"
>      ],
>      "unlocked_by": null,
>      "uuid": "d74be875-bc1c-47a5-bf80-0c4c57ed200e"
>    }

To just about the entry for armstrong numbers

**Comment by jvarness** [comment]
I've moved the exercise above the armstrong numbers exercise. Let me know if there's anything else you'd like for me to do.

**Pull request merged**